doc: create ADR 0003 for Skate MQTT (#2685)
mbta/skate: documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md

# 3. Use MQTT to publish Trip Modifications

Date: 2024-07-04

## Status

Proposed

## Context

We need to send [GTFS Trip Modifications](https://gtfs.org/realtime/feed-entities/trip-modifications/)
to Transit Data.

## Decision

To minimize the amount of resources used and latency between Skate publishing or
un-publishing a Trip Modification, we are going to use MQTT to form a 

> Event Stream between Skate and Transit Data.

Reviewer: ```suggestion
event stream between Skate and Transit Data.
```

Should "event stream" be capitalized?

Since this is an output of Skate, we will not make this service required to do
local development.

## Consequences

> We'll be adding a new technology and service to Skate, this is more surface area

Reviewer: ```suggestion
We'll be adding a new technology and service to Skate, so there will be more surface area
```

OR

```suggestion
Since we'll be adding a new technology and service to Skate, there will be more surface area
```

Tiny grammar nit.

> to test.

Reviewer: Are there any additional risks that we need to consider?

Here are a few possible ones:
1. We need to be super careful in order to ensure that Transit Data gets up-to-date detour info. If the wrong message winds up being retained, then Transit Data won't have any way to get the actual latest message.
   - If we run purely on fire-and-forget, then if a message doesn't make it to the MQTT broker for whatever reason, then it never will. 
   - If we do retry sending messages on failure, then we could wind up sending messages out of order.
1. We may need to keep stale topics around for a long time so that TD knows that a detour has been deactivated.

Author: > If the wrong message winds up being retained

I don't envision this as being an issue, can you elaborate on why it would be?

Also, I don't believe this risk would belong in this ADR either.

Author: > We may need to keep stale topics around for a long time so that TD knows that a detour has been deactivated.

I think this is also not a concern for this ADR. But I also don't think we need to keep stale topics around, we can discuss further in #2686 if you'd like.

Reviewer: > I don't envision this as being an issue, can you elaborate on why it would be?

If we're not careful (and implement retry in the wrong way), we could wind up with a sequence of events like:
- Create Detour A (communicated with `message 1`)
- Sending `message 1` fails - queue a retry
- Update Detour A in some way (communicated with `message 2`)
- Sending `message 2` succeeds
- Sending `message 1` retry succeeds

Now the latest message about Detour A is `message 1`, which doesn't have whatever the updated info from `message 2` is.

Making sure that the right message is retained is a solvable problem, but any time messages are a thing, there is that bit of risk that messages will get delivered out of order.

Author: > If we're not careful (and implement retry in the wrong way),

What do you think about this?
```diff
diff --git a/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md b/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md
index f72ba5caf2...8cb8b8c8bc 100644
--- a/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md
+++ b/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md
@@ -57,6 +57,10 @@
 Since we'll be adding a new technology and service to Skate, there will be more
 surface area to test.
 
+By using a message queue, we risk sending events out of order and
+having inaccurate ordering of data, we'll need to be careful to address these
+issues as they arise during implementation.
+
 [^ps-1]: In this comment [Paul Swartz mentioned that using MQTT but sending all
   messages using `retain` will effectively be the same as uploading to S3, but
   with lower latency to clients](https://app.asana.com/0/0/1207124404105396/1207386033651835/f).
```

# 3. Use MQTT to publish Trip Modifications

Date: 2024-07-04

## Status

Proposed

## Context

We need to send [GTFS Trip Modifications](https://gtfs.org/realtime/feed-entities/trip-modifications/)
to Transit Data.

## Decision

To minimize the amount of resources used and latency between Skate publishing or
un-publishing a Trip Modification and Transit Data becoming aware, we are going
to use MQTT to form an event stream between Skate and Transit Data.

Since this is an output of Skate, we will not make this service required to do
local development.

### Alternatives Considered

#### Polling based alternatives
1. HTTP API

>   - We _could_ host an internal API on Skate which Transit Data can Poll. A
>   benefit of this is we can "render straight from the database" and avoid any
>   sort of expiration/synchronization logic with an external service.
>   Additionally, it is already an established practice within Transit Data to
>   poll an API (Ex: `Concentrate` polls `RTR`).  
>   The downside being that we are then requiring extra resources to continuously
>   poll, at 1s intervals, to check for any changes. Additionally, Skate's current
>   access control mechanism is concerned entirely with _users_ and not providing
>   pragmatic access. Using an external service allows us to push authentication
>   for this access to other services and IaC instead of requiring that Skate also
>   concerns itself with these issues.

Reviewer: Another worthwhile point is that an API would be a good choice if "reads" were infrequent while "writes" were frequent, but the reality is the other way around - "reads" happen once a second, or any time a user wants to look something up, while write will only happen a few times a week

Some possible wordsmithing (no need to take this suggestion exactly):

```suggestion
  - We considered delivering this information by way of an API that
    Transit Data could poll. That would have the benefit of allowing
    us to "render straight from the database" and avoid any sort of
    expiration/synchronization logic with an external service.
    Additionally, it is already an established practice within Transit
    Data to poll an API (Ex: `Concentrate` polls `RTR`).

    This might be an appropriate choice if we expected there to be
    more "writes" than "reads" -- that is, if we expected officials to
    be modifying detours more frequently than Transit Data would be
    querying them. But we actually expect many more "reads" than
    "writes" - we'll need to access this data any time a rider looks
    up a bus route, or any time an operator loads turn-by-turn
    directions, or possibly once per second, while writes will only
    happen a few times a week, when an official is creating or
    modifying a detour. Creating an API would create performance
    concerns that messaging wouldn't.

    Additionally, Skate's current access control mechanism is
    concerned entirely with _users_ and not providing pragmatic
    access. Using an external service allows us to push authentication
    for this access to other services and IaC instead of requiring
    that Skate also concerns itself with these issues.
```

Rendered:
>  - We considered delivering this information by way of an API that
    Transit Data could poll. That would have the benefit of allowing
    us to "render straight from the database" and avoid any sort of
    expiration/synchronization logic with an external service.
    Additionally, it is already an established practice within Transit
    Data to poll an API (Ex: `Concentrate` polls `RTR`).
>
>    This might be an appropriate choice if we expected there to be
    more "writes" than "reads" -- that is, if we expected officials to
    be modifying detours more frequently than Transit Data would be
    querying them. But we actually expect many more "reads" than
    "writes" - we'll need to access this data any time a rider looks
    up a bus route, or any time an operator loads turn-by-turn
    directions, or possibly once per second, while writes will only
    happen a few times a week, when an official is creating or
    modifying a detour. Creating an API would create performance
    concerns that messaging wouldn't.
>
>    Additionally, Skate's current access control mechanism is
    concerned entirely with _users_ and not providing pragmatic
    access. Using an external service allows us to push authentication
    for this access to other services and IaC instead of requiring
    that Skate also concerns itself with these issues.
